### PR TITLE
3. Split up Translator and SchemaBuilder

### PR DESCRIPTION
**Pull request by Andy2003**
Merge after #48 and #49

### CODE REVIEW

File: src/main/kotlin/org/neo4j/graphql/Cypher.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+import graphql.schema.GraphQLFieldDefinition
+import graphql.schema.GraphQLObjectType
+import graphql.schema.GraphQLType
+
+data class Cypher(val query: String, val params: Map<String, Any?> = emptyMap()) {
+    fun with(p: Map<String, Any?>) = this.copy(params = this.params + p)
+    fun escapedQuery() = query.replace("\"", "\\\"").replace("'", "\\'")
+
+    companion object {
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/Cypher.kt]: I would move these somewhere else don't really belong into this data class imho.
It looks almost a bit like the DSL bits we discussed.

Comment (Andy2003) [reply src/main/kotlin/org/neo4j/graphql/Cypher.kt]: already removed in the 2nd commit, the 1st commit is just to have an idea about the refactoring

**Comment by jexp** [comment]
Has also conflicts and needs to be rebased

**Comment by jexp** [comment]
I checked the hierarchies in graphql-java again, and it seems that GraphQLFieldsContainer
would be what we want as supertype of interface and objec type,

so that means we could most likely replace NodeFacade (we need a better name for that anyway) with it.

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            wireDefinedOperations(queryDefinition, true)
+            wireDefinedOperations(mutationDefinition, false)
+
+            val relationTypes = objectTypeDefinitions
+                .filter { it.getDirective(DirectiveConstants.RELATION) != null }
+                .map { it.getDirective(DirectiveConstants.RELATION).getArgument(DirectiveConstants.RELATION_NAME).value.toJavaValue().toString() to it }
+                .toMap()
+
+            val nodeDefinitions: List<TypeDefinition<*>> = interfaceTypeDefinitions + objectTypeDefinitions
+            nodeDefinitions.forEach { createNodeMutation(it.getNodeType()!!) }
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt]: what does it mean to create an interface?

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+        /**
+         * add the given operation to the corresponding ObjectTypeDefinition
+         */
+        private fun mergeOperation(objectTypeDefinition: ObjectTypeDefinition, dataFetcher: BaseDataFetcher?)
+                : ObjectTypeDefinition {
+            if (dataFetcher == null) {
+                return objectTypeDefinition
+            }
+            if (objectTypeDefinition.fieldDefinitions.any { fd -> dataFetcher.fieldDefinition.name == fd.name }) {
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt]: what if only some have one but others don't?

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+                    objectTypeDefinitions.firstOrNull { it.name == targetField.type.inner().name() }
+                        ?.let { target ->
+                            mutationDefinition = mergeOperation(mutationDefinition, DeleteRelationHandler.build(source, target, metaProvider))
+                            mutationDefinition = mergeOperation(mutationDefinition, CreateRelationHandler.build(source, target, relationTypes, metaProvider))
+                        }
+                }
+        }
+
+
+        private fun <T : SDLDefinition<*>> enhanceRelations(source: T, childName: String): T {
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt]: todo better function name

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+            if (schemaConfig.mutation.enabled && !schemaConfig.mutation.exclude.contains(typeName)) {
+                if (type is ObjectDefinitionNodeFacade) {
+                    mutationDefinition = if (type.isRelationType()) {
+                        mergeOperation(mutationDefinition, CreateRelationTypeHandler.build(type, metaProvider))
+                    } else {
+                        mergeOperation(mutationDefinition, CreateTypeHandler.build(type, metaProvider))
+                    }
+                }
+                mutationDefinition = mergeOperation(mutationDefinition, DeleteHandler.build(type, metaProvider))
+                mutationDefinition = mergeOperation(mutationDefinition, MergeOrUpdateHandler.build(type, true, metaProvider))
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt]: TODO later annotate true/false with param name

File: src/main/kotlin/org/neo4j/graphql/handler/BaseDataFetcher.kt [hunk 6]
```diff
@@ -0,0 +1,149 @@
+package org.neo4j.graphql.handler
+
+import graphql.language.*
+import graphql.schema.DataFetcher
+import graphql.schema.DataFetchingEnvironment
+import org.neo4j.graphql.*
+import org.neo4j.graphql.handler.projection.ProjectionBase
+
+abstract class BaseDataFetcher(
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/BaseDataFetcher.kt]: todo later rename to CypherQueryFetcher

File: src/main/kotlin/org/neo4j/graphql/handler/BaseDataFetcher.kt [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    fun label(includeAll: Boolean = false) = type.label(includeAll)
+
+
+    protected fun properties(variable: String, arguments: List<Argument>): Cypher {
+        val all = preparePredicateArguments(arguments)
+        if (all.isEmpty()) {
+            return Cypher.EMPTY
+        }
+        val query = all
+            .joinToString(", ", " { ", " }") { it.toCypherString(variable) }
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/BaseDataFetcher.kt]: later: JTS

File: src/main/kotlin/org/neo4j/graphql/handler/BaseDataFetcher.kt [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                idProperty: Argument?,
+                idField: FieldDefinition,
+                paramName: String? = idProperty?.let { paramName(variable, idProperty.name, idProperty.value) }
+        ): Cypher {
+            return when {
+                idProperty != null && paramName != null -> {
+                    val queryParams = mapOf(paramName to idProperty.value.toJavaValue())
+                    // TODO handle @property aliasing
+                    if (idProperty.value is ArrayValue) {
+                        Cypher("($variable:$label) WHERE  $variable.${idField.name.quote()} IN $$paramName", queryParams)
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/BaseDataFetcher.kt]: todo quote labels

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt (hunk 2 above)
Comment (Andy2003) [reply src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt]: You can update / merge / delete interfaces , but not create them

File: src/main/kotlin/org/neo4j/graphql/handler/CypherDirectiveHandler.kt [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    : BaseDataFetcher(type, fieldDefinition, metaProvider) {
+
+    override fun generateCypher(variable: String, field: Field, projectionProvider: () -> Cypher, env: DataFetchingEnvironment): Cypher {
+        val mapProjection = projectionProvider.invoke()
+        val ordering = orderBy(variable, field.arguments)
+        val skipLimit = format(skipLimit(field.arguments))
+
+        return if (isQuery) {
+            val (query, params) = cypherDirective(variable, fieldDefinition, field, cypherDirective, emptyList())
+            Cypher("UNWIND $query AS $variable" +
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/CypherDirectiveHandler.kt]: later: we should make this consistent with the one below

File: src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt (hunk 3 above)
Comment (Andy2003) [reply src/main/kotlin/org/neo4j/graphql/SchemaBuilder.kt]: this method is called per field on a root definition (mutation / query). If a dataFetcher was already defined we skip this field.

File: src/main/kotlin/org/neo4j/graphql/handler/MergeOrUpdateHandler.kt [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        type: NodeFacade,
+        private val merge: Boolean,
+        private val idField: FieldDefinition,
+        fieldDefinition: FieldDefinition,
+        metaProvider: MetaProvider
+) : BaseDataFetcher(type, fieldDefinition, metaProvider) {
+
+    companion object {
+        fun build(type: NodeFacade, merge: Boolean, metaProvider: MetaProvider): MergeOrUpdateHandler? {
+            val idField = type.fieldDefinitions().find { it.isID() } ?: return null
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/MergeOrUpdateHandler.kt]: later: need to handle-mult-id fields

File: src/main/kotlin/org/neo4j/graphql/handler/QueryHandler.kt [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    override fun generateCypher(variable: String, field: Field, projectionProvider: () -> Cypher, env: DataFetchingEnvironment): Cypher {
+
+        val mapProjection = projectionProvider.invoke()
+        val ordering = orderBy(variable, field.arguments)
+        val skipLimit = format(skipLimit(field.arguments))
+
+        val select = if (type.isRelationType()) {
+            "()-[$variable:${label()}]->()"
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/QueryHandler.kt]: quoting

File: src/main/kotlin/org/neo4j/graphql/handler/QueryHandler.kt [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        } else {
+            "($variable:${label()})"
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/QueryHandler.kt]: quoting

File: src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+                        .filterIsInstance<Field>()
+                        .filter { it.name == neo4jType.name || it.alias == neo4jType.name }
+                }
+                .groupBy({ it.first }, { it.second }) // create a map of <FieldWithNeo4jType, List<Field>> so we group the data by the type
+                .mapValues { it.value.flatten() }
+                .flatMap { entry ->
+                    // for each FieldWithNeo4jType of type query we create the where condition
+                    val typeName = entry.key.type.name()
+                    val fields = entry.value
+                    val neo4jType = metaProvider.getNodeType(typeName)
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt]: perhaps we should encapsulate that exception throwing if it's a common case

Comment (jexp) [reply src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt]: for getType()

File: src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+                                    .forQuery(arg, f, neo4jType)
+                                Translator.CypherArgument(nameSuffix, propertyNameSuffix, arg.value, innerNeo4jConstruct)
+                            }
+                    }
+                }
+
+    private fun argumentsToMap(arguments: List<Argument>, resultObjectType: NodeFacade? = null): Map<String, Translator.CypherArgument> {
+        return arguments
+            .map { argument ->
+                val propertyName = (resultObjectType?.getFieldDefinition(argument.name)?.propertyDirectiveName()
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt]: later: extract function or such

Comment (jexp) [reply src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt]: for @property directtive aliases

File: src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+        } else emptyList()
+    }
+
+    fun propertyArguments(queryField: Field) =
+            queryField.arguments.filterNot { listOf(FIRST, OFFSET, ORDER_BY).contains(it.name) }
+
+    private fun preparePredicateArguments(field: FieldDefinition, arguments: List<Argument>): List<Translator.CypherArgument> {
+        if (arguments.isEmpty()) return emptyList()
+        val resultObjectType = metaProvider.getNodeType(field.type.name())
+                ?: throw IllegalArgumentException("${field.name} cannot be converted to a NodeGraphQlFacade")
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt]: no implementation detail exceptions

File: src/main/kotlin/org/neo4j/graphql/handler/QueryHandler.kt (hunk 11 above)
Comment (Andy2003) [reply src/main/kotlin/org/neo4j/graphql/handler/QueryHandler.kt]: label() is already quoting

File: src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+        val fieldProjection = projectFields(childVariable, field, fieldObjectType, env, variableSuffix)
+
+        val comprehension = "[$childVariable IN ${expression.query} | ${fieldProjection.query}]"
+        val skipLimit = skipLimit(field.arguments)
+        val slice = slice(skipLimit, fieldType.isList())
+        return Cypher(comprehension + slice, (expression.params + fieldProjection.params)) // + where.params
+
+    }
+
+    private fun relationshipInfoInCorrectDirection(fieldObjectType: NodeFacade, relInfo0: RelationshipInfo, parent: NodeFacade, relDirectiveField: RelationshipInfo?): RelationshipInfo {
```
Comment (jexp) [inline src/main/kotlin/org/neo4j/graphql/handler/projection/ProjectionBase.kt]: later: I want . to extract and clean up all that relationship-handling

File: src/test/resources/movie-tests.adoc [hunk 17]
```diff
@@ -650,12 +650,11 @@ mutation {
...
-MATCH (from:Movie { movieId: $movieId })
-MATCH (to:Genre)
-WHERE to.name IN $genres
-MERGE (from)-[r:IN_GENRE]->(to)
-WITH DISTINCT from
-RETURN from { .title } AS movie
+MATCH (from:Movie { movieId: $fromMovieId })
+MATCH (to:Genre) WHERE to.name IN $toGenres
+MERGE (from)-[:IN_GENRE]->(to)
+WITH DISTINCT from AS add
```
Comment (jexp) [inline src/test/resources/movie-tests.adoc]: the name should probably not be add ?

**Pull request merged**